### PR TITLE
adds rewards address linking and rewards for pending supply order

### PR DESCRIPTION
**Pull request by jennypollack**
fixes token precision price
removes system addresses from rewards logic
deployment pending 
[kovan](https://thegraph.com/explorer/subgraph/jennypollack/tinlake-rewards-staging)
[mainnet](https://thegraph.com/explorer/subgraph/jennypollack/tinlake-rewards)

### CODE REVIEW

File: src/domain/Account.ts [hunk 1]
```diff
@@ -10,22 +12,26 @@ export function createAccount(address: string): Account {
 }
 
 // used for determining nonZeroBalanceSince across system for user
-export function updateAccounts(event: TransferEvent): void {
-  // increase accountTo balance
-  let accountTo = Account.load(event.params.dst.toHex())
-  if (accountTo == null) {
-    accountTo = createAccount(event.params.dst.toHex())
+export function updateAccounts(event: TransferEvent, poolId: string): void {
```
Comment (philipstanislaus) [inline src/domain/Account.ts:15]: Can we give this function a name that better resembles the tying to the transfer event, e. g. `updateAccountsAfterTransfer`?

File: src/domain/Reward.ts [hunk 2]
```diff
@@ -28,8 +29,7 @@ export function loadOrCreateRewardDayTotal(date: BigInt): RewardDayTotal {
     rewardDayTotal = new RewardDayTotal(date.toString())
     rewardDayTotal.todayValue = BigInt.fromI32(0)
     rewardDayTotal.toDateAggregateValue = BigInt.fromI32(0)
-    // 0.0042 RAD/DAI up to 1M RAD
-    rewardDayTotal.rewardRate = BigDecimal.fromString(initialRewardRate)
+    rewardDayTotal.rewardRate = BigDecimal.fromString('0')
```
Comment (philipstanislaus) [inline src/domain/Reward.ts:32]: Why set the rate to 0?

File: src/domain/Reward.ts [hunk 3]
```diff
@@ -125,12 +143,12 @@ export function calculateRewards(date: BigInt, pool: Pool): void {
   systemRewards.save()
 }
 
-function checkRewardRate(checker: RewardDayTotal): void {
-  if (checker.toDateRewardAggregateValue.gt(BigDecimal.fromString(tierOneRewards))) {
-    log.debug('pending rewards > 1 MILL:  {}', [checker.toDateRewardAggregateValue.toString()])
+function setRewardRate(systemRewards: RewardDayTotal): RewardDayTotal {
```
Comment (philipstanislaus) [inline src/domain/Reward.ts:146]:
```suggestion
function updateRewardRate(systemRewards: RewardDayTotal): RewardDayTotal {
```

File: src/domain/Reward.ts [hunk 4]
```diff
@@ -125,12 +143,12 @@ export function calculateRewards(date: BigInt, pool: Pool): void {
...
+  if (systemRewards.toDateRewardAggregateValue.lt(BigDecimal.fromString(tierOneRewards))) {
```
Comment (philipstanislaus) [inline src/domain/Reward.ts:147]: No need to update if it's already correct:

```suggestion
  if (systemRewards.toDateRewardAggregateValue.lt(BigDecimal.fromString(tierOneRewards)) && systemRewards.rewardRate.neq(BigDecimal.fromString('0.0042')) {
```

File: src/domain/Reward.ts [hunk 5]
```diff
@@ -125,12 +143,12 @@ export function calculateRewards(date: BigInt, pool: Pool): void {
...
+    log.debug('setting system rewards rate {}', [systemRewards.toDateRewardAggregateValue.toString()])
 
-    // TODO: update this with correct value
-    checker.rewardRate = BigDecimal.fromString('0')
-    checker.save()
+    systemRewards.rewardRate = BigDecimal.fromString('0.0042')
+    systemRewards.save()
   }
```
Comment (philipstanislaus) [inline src/domain/Reward.ts:152]: Don't we need an else case here to set reward rate to `0` or another value after tier one?

File: src/domain/Reward.ts [hunk 6]
```diff
@@ -125,12 +143,12 @@ export function calculateRewards(date: BigInt, pool: Pool): void {
...
-  if (checker.toDateRewardAggregateValue.gt(BigDecimal.fromString(tierOneRewards))) {
-    log.debug('pending rewards > 1 MILL:  {}', [checker.toDateRewardAggregateValue.toString()])
+function setRewardRate(systemRewards: RewardDayTotal): RewardDayTotal {
+  if (systemRewards.toDateRewardAggregateValue.lt(BigDecimal.fromString(tierOneRewards))) {
+    log.debug('setting system rewards rate {}', [systemRewards.toDateRewardAggregateValue.toString()])
 
-    // TODO: update this with correct value
-    checker.rewardRate = BigDecimal.fromString('0')
-    checker.save()
+    systemRewards.rewardRate = BigDecimal.fromString('0.0042')
```
Comment (philipstanislaus) [inline src/domain/Reward.ts:150]: Can we put the rate here into the config?

**Review by philipstanislaus — APPROVED** [review]
LGTM, thanks! 🎉 

One generic comment: The subgraph gets more and more complex, I feel we should really get started with some functional tests that spin up a docker based graph node, deploy tinlake to tinlake-in-a-box, deploy our subgraph, and then run a whole bunch of tests that test edge cases such as changes of reward claim addresses, moving DROP/TIN tokens and then claiming etc...

I just added a ticket for that: #94

File: schema.graphql [hunk 7]
```diff
@@ -99,19 +99,24 @@ type ERC20Transfer @entity {
   pool: Pool!
 }
 
+# used in reward calc
```
[the remark this answers is not in the document]
Comment (lucasvo) [reply schema.graphql:102]: I don't think this adds to the readability of the code.

File: src/domain/Reward.ts (hunk 2 above)
Comment (lucasvo) [reply src/domain/Reward.ts:32]: We will likely update the logic by which we set the reward rate over time thus initializing this with the correct value would lead to code for the reward logic to be scattered in the code-base. In the future the reward logic will look something like this:

```
if (block < 1000) { 
    // initial reward logic
} else if (block < 2000) { 
   // updated reward logic
} else if (block < 3000)
  // more updates
} else {
   // current reward logic
}
```

File: src/domain/Reward.ts (hunk 5 above)
Comment (lucasvo) [reply src/domain/Reward.ts:152]: No because the default value is 0

File: src/mappings/TinlakeClaimRAD.ts [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+  let balance = loadOrCreateRewardBalance(sender)
+  let link = loadOrCreateRewardLink(sender, substrateAccount)
+  balance.claims = push(balance.claims, link.id)
+
+  // add this link to their reward balance and put any
+  // eligible rewards claimable into this claim, reset claimable to 0
+  if (balance.eligible) {
+    link.rewardsAccumulated = balance.claimableRewards
+    balance.claimableRewards = BigDecimal.fromString('0')
+  }
```
[the remark this answers is not in the document]
Comment (lucasvo) [reply src/mappings/TinlakeClaimRAD.ts:23]: Yes this is intended. The Merkle tree should be populated from `RewardLink`

File: src/domain/RewardLink.ts [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+import { RewardLink } from '../../generated/schema'
+
+export function loadOrCreateRewardLink(address: string, substrate: string): RewardLink {
+  let claim = RewardLink.load(address.concat(substrate))
+  if (claim == null) {
+    new RewardLink(address.concat(substrate))
+    claim.ethAddress = address
+    claim.centAddress = substrate
+    claim.rewardsAccumulated = BigDecimal.fromString('0')
+    claim.save()
```
[the remark this answers is not in the document]
Comment (jennypollack) [reply src/domain/RewardLink.ts:11]: good catch, adding ```claim =``` new RewardLink()

File: subgraph-kovan-staging.yaml [hunk 10]
```diff
@@ -205,4 +207,42 @@ templates:
...
+      entities:
+        - TokenBalance
+      abis:
+        - name: Tranche
+          file: ./abis/Tranche.json
+        - name: ERC20
+          file: ./abis/ERC20.json
+      callHandlers:
+        - function: supplyOrder(address,uint256)
+          handler: handleSupplyOrder
```
[the remark this answers is not in the document]
Comment (lucasvo) [reply subgraph-kovan-staging.yaml:229]: We don't need to handle redeem because upon placing a redeem order you move the ERC20 tokens into the Tranche thus they are already not counted anymore when we look at token balances.

File: src/domain/Reward.ts (hunk 6 above)
Comment (lucasvo) [reply src/domain/Reward.ts:150]: No this is not a config option and shouldn't be changed ever. Similarly to my previous comment (see https://github.com/centrifuge/tinlake-subgraph/pull/93/files#r550752175) the goal is to keep all reward logic in one file where new reward logic is appended every time we update it.

File: src/mappings/Tranche.ts [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+  let token = tokenAddresses.juniorToken
+  if (tokenAddresses.seniorTranche == tranche) {
+    token = tokenAddresses.seniorToken
+  }
+
+  let account = call.inputs.usr.toHex()
+  let amount = call.inputs.newSupplyAmount
+
+  // protection from adding system account to internal tracking
+  if (!isSystemAccount(poolId, account)) {
```
[the remark this answers is not in the document]
Comment (jennypollack) [reply src/mappings/Tranche.ts:24]: 👍

File: src/mappings/Tranche.ts [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+      createAccount(account)
+    }
+    addToGlobalAccounts(account)
+    // add to token owners
+    let tk = loadOrCreateToken(token)
+    tk.owners = push(tk.owners, account)
+    tk.save()
+
+    let tokenBalance = loadOrCreateTokenBalance(account, token)
+    tokenBalance.pendingSupplyCurrency = amount
```
[the remark this answers is not in the document]
Comment (jennypollack) [reply src/mappings/Tranche.ts:35]: Tinlake's supplyOrder method takes the total supply amount as an arg, so overwriting here is correct

File: src/domain/Reward.ts (hunk 6 above)
Comment (lucasvo) [inline src/domain/Reward.ts:150]: @philipstanislaus this value is actually not correct as we need to multiply this by the number of decimals we have for RAD in substrate. It should be `0.0042*10**18` right?

Comment (philipstanislaus) [reply src/domain/Reward.ts:150]: I have to double check, but on first glance 0.0042 sounds correct, since it should be a fraction.

Comment (jennypollack) [reply src/domain/Reward.ts:150]: we checked this yesterday and it's correct. also working on some scripts to validate rewards amounts separately which are looking good 👍 still need to do the test claims on kovan and mainnet to fully validate

File: src/domain/Day.ts [hunk 13]
```diff
@@ -30,6 +30,9 @@ export function getToday(block: ethereum.Block): Day {
 
 // if the difference between days since nonzerobalance
 // and today's timestamp are greater than or equal to sixty days in seconds
-export function haveSixtyDaysPassed(today: BigInt, nonZeroSince: BigInt): boolean {
-  return nonZeroSince.minus(BigInt.fromI32(secondsInSixtyDays)) >= today
+// if kovan, we want to make rewards eligible after 1 day
+export function rewardsAreEligible(today: BigInt, nonZeroSince: BigInt): boolean {
+  return dataSource.network() == 'mainnet'
+    ? nonZeroSince.minus(BigInt.fromI32(secondsInSixtyDays)) >= today
+    : nonZeroSince.minus(BigInt.fromI32(secondsInDay)) >= today
```
Comment (philipstanislaus) [inline src/domain/Day.ts:37]: Should this not be: 

```suggestion
    ? nonZeroSince.plus(BigInt.fromI32(secondsInSixtyDays)) >= today
    : nonZeroSince.plus(BigInt.fromI32(secondsInDay)) >= today
```

File: src/domain/Day.ts [hunk 14]
```diff
@@ -30,9 +30,9 @@ export function getToday(block: ethereum.Block): Day {
...
-// if kovan, we want to make rewards eligible after 1 day
-export function rewardsAreEligible(today: BigInt, nonZeroSince: BigInt): boolean {
-  return dataSource.network() == 'mainnet'
-    ? nonZeroSince.minus(BigInt.fromI32(secondsInSixtyDays)) >= today
-    : nonZeroSince.minus(BigInt.fromI32(secondsInDay)) >= today
+// if kovan, we want to make rewards claimable after 1 day
+export function rewardsAreClaimable(today: BigInt, nonZeroSince: BigInt | null): boolean {
+  return nonZeroSince != null && dataSource.network() == 'mainnet'
+    ? today.minus(<BigInt>nonZeroSince).ge(BigInt.fromI32(secondsInSixtyDays))
+    : today.minus(<BigInt>nonZeroSince).ge(BigInt.fromI32(secondsInDay))
```
Comment (philipstanislaus) [inline src/domain/Day.ts:37]: There are parentheses missing here: 

```suggestion
  return nonZeroSince != null && (dataSource.network() == 'mainnet'
    ? today.minus(<BigInt>nonZeroSince).ge(BigInt.fromI32(secondsInSixtyDays))
    : today.minus(<BigInt>nonZeroSince).ge(BigInt.fromI32(secondsInDay)))
```

**Review by philipstanislaus — CHANGES_REQUESTED** [review]
Nice, only one thing stood out to me, then we can merge this!

**Review by philipstanislaus — APPROVED** [review]
Awesome, let's get this merged 🙌 🎉
